### ISSUE TITLE
Add Apple MPS support for GPU

### ISSUE THREAD

**Issue by dogjamboree** [Lex-au/Orpheus-FastAPI#17]
Please add MPS support for Apple MPS for those running Apple silicon.

**Comment by Lex-au** [comment]
Hi @dogjamboree 

Thanks for the suggestion. I'm sorry to hear it's not working natively for you. Can you confirm what version of Pytorch you're running? I infer that due to my direct preference for CUDA, you're being forced to run inference by CPU?

Could you dump some terminal logs for me?

**Comment by LucidScript** [comment]
Hey @Lex-au I used Pinokio to instal your "[Orpheus-FastAPI]" on mac silicon :D and it is working ( :. Well it is only CPU, but  still great for now :D thank you for your hard work ( :. I hope in future we will have metal support. 

And I have 2 off-topic question :)
1. Is it possible to add pauses inside the conversation in seconds?
2. Is there a way to produce more than 2 minutes audio file?

Emm thank you again :D, I am really happy that I can test this one mac silicon :D

**Comment by CicadaDigital** [comment]
+1 for MPS support and optimization would go a long way. Thank you for doing gods work.

**Comment by dogjamboree** [comment]
> Hi [@dogjamboree](https://github.com/dogjamboree)
> 
> Thanks for the suggestion. I'm sorry to hear it's not working natively for you. Can you confirm what version of Pytorch you're running? I infer that due to my direct preference for CUDA, you're being forced to run inference by CPU?
> 
> Could you dump some terminal logs for me?

I should have been clearer. I didn't attempt an install but was preemptively requesting MPS support since I figured CPU would run slower than native MPS. I will give it a shot but it does sound like it's working on CPU from others' responses. Thanks!

**Comment by Lex-au** [comment]
> Hey [@Lex-au](https://github.com/Lex-au) I used Pinokio to instal your "[Orpheus-FastAPI]" on mac silicon :D and it is working ( :. Well it is only CPU, but still great for now :D thank you for your hard work ( :. I hope in future we will have metal support.
> 
> And I have 2 off-topic question :)
> 
> 1. Is it possible to add pauses inside the conversation in seconds?
> 2. Is there a way to produce more than 2 minutes audio file?
> 
> Emm thank you again :D, I am really happy that I can test this one mac silicon :D

Hi! 

Really appreciate the kind words—and super cool that you're running it on Mac Silicon through Pinokio! I knew another user here mentioned it had Orpheus-FastAPI there, didn't realise it was getting big though... but I’m glad it works out of the box (even if just CPU for now). Metal support would be amazing, and I’m definitely keen to explore it if/when PyTorch adds more native backing for Metal.

To your questions:

1. Pauses in conversation (in seconds)?
Yep! You can simulate pauses by adding punctuation like "..." or just breaking up text with full stops and newlines. If you're using an LLM twith Orpheus as as the speech endpoint, then you will need to feed your LLM a system prompt. I included one on the git, but you can fine-tune it to also include more "...".

2. More than 2 minutes of audio?
Totally possible! In v1.1, Orpheus splits input text every 1000 characters and stitches the audio back together using cross-fade logic. You can feed in as much text as you'd like (tts.html you can edit the free text box from 8192 characters to anything you want)—just be aware that longer outputs will include tiny stitching artifacts. It's a small trade-off, but I've done my best to work-around the underlying models constraints.

Example: Half of chapter 1 of Moby Dick.

![Image](https://github.com/user-attachments/assets/70cb54d7-9732-4491-9433-6e353a1e8ab1)

Thanks again for testing and for the positive energy—it means a lot!

**Comment by dogjamboree** [comment]
I don't understand why this issue was closed. Despite the fact that the code will run as-is using cpu, it looks like mps optimizations could still be made. I'm not a coder and and I don't mean any disrespect so please take everything I say with a grain of salt but based on a perplexity / Claude 3.7 search I did of your inference.py script the suggestions below were given (see text below separator). 

Would it not be prudent to leave to issue open until someone with apple hardware and programming knowledge could take a look and see if they can optimize the code? Or am I being naive? (I honestly have no idea!). 

Thanks again for your hard work and again, no disrespect meant with these tickets.

---------------------------------------

The code you've shared is a Python module for text-to-speech synthesis using the Orpheus model, which appears to be designed to run on various hardware configurations including GPUs. To optimize this code specifically for Apple's Metal Performance Shaders (MPS), I'll focus on key areas that could benefit from MPS acceleration.

## Hardware Detection and Optimization

The code already has hardware detection logic, but it's focused on CUDA GPUs. To optimize for Apple MPS:

1. Add MPS detection alongside CUDA detection:

```python
# After the existing CUDA detection code
MPS_AVAILABLE = False
if not torch.cuda.is_available() and hasattr(torch, 'mps') and torch.mps.is_available():
    MPS_AVAILABLE = True
    mps_device = torch.device("mps")
    print(f"🖥️ Hardware: Apple Silicon GPU detected")
    print(f"🚀 Using MPS-optimized settings")
```

2. Modify the HIGH_END_GPU detection to include high-end Apple Silicon:

```python
# Consider M1 Pro/Max/Ultra or M2 Pro/Max/Ultra as high-end
if hasattr(torch, 'mps') and torch.mps.is_available():
    import platform
    model = platform.processor()
    HIGH_END_GPU = any(x in model for x in ["M1 Pro", "M1 Max", "M1 Ultra", "M2 Pro", "M2 Max", "M2 Ultra", "M3 Pro", "M3 Max"])
    if HIGH_END_GPU:
        print(f"🖥️ Hardware: High-end Apple Silicon detected: {model}")
        print("🚀 Using high-performance optimizations")
```

## Tensor Operations Optimization

For any tensor operations in the code, ensure they're running on the MPS device:

```python
def convert_to_audio(multiframe: List[int], count: int) -> Optional[bytes]:
    """Convert token frames to audio with MPS acceleration when available."""
    # Import here to avoid circular imports
    from .speechpipe import convert_to_audio as orpheus_convert_to_audio
    
    # If using numpy arrays in the conversion, move to MPS if available
    if MPS_AVAILABLE:
        # Example: If there's numpy conversion that could be accelerated
        # Convert numpy operations to torch operations on MPS device
        start_time = time.time()
        result = orpheus_convert_to_audio(multiframe, count, use_mps=MPS_AVAILABLE)
    else:
        start_time = time.time()
        result = orpheus_convert_to_audio(multiframe, count)
    
    if result is not None:
        perf_monitor.add_audio_chunk()
        
    return result
```

## Memory Management for MPS

Apple Silicon benefits from unified memory, but still requires careful memory management:

```python
# Add to the tokens_decoder_sync function
def tokens_decoder_sync(syn_token_gen, output_file=None):
    # Existing code...
    
    # For MPS, optimize memory usage with smaller batches and more frequent processing
    if MPS_AVAILABLE:
        # MPS benefits from more frequent, smaller batches
        batch_size = 16  # Smaller batch size for MPS
        # Enable attention slicing for MPS (if applicable)
        use_attention_slicing = True
    else:
        batch_size = 32 if HIGH_END_GPU else 16
        use_attention_slicing = False
    
    # Rest of the function...
```

## Batch Processing Optimization

For the batch processing in `generate_speech_from_api`, adjust batch sizes for MPS:

```python
def generate_speech_from_api(prompt, voice=DEFAULT_VOICE, output_file=None, temperature=TEMPERATURE, 
                     top_p=TOP_P, max_tokens=MAX_TOKENS, repetition_penalty=None, 
                     use_batching=True, max_batch_chars=1000):
    # Existing code...
    
    # Adjust batch size for MPS
    if MPS_AVAILABLE:
        # MPS often performs better with smaller batch sizes
        max_batch_chars = 800 if not HIGH_END_GPU else 1000
    
    # Rest of the function...
```

## Audio Processing Optimization

For audio processing with NumPy, consider using torch operations on MPS:

```python
def stream_audio(audio_buffer):
    """Stream audio buffer to output device with error handling and MPS optimization."""
    if audio_buffer is None or len(audio_buffer) == 0:
        return
    
    try:
        if MPS_AVAILABLE:
            # Convert to torch tensor and process on MPS
            audio_tensor = torch.frombuffer(audio_buffer, dtype=torch.int16).to(mps_device)
            audio_float = audio_tensor.float() / 32767.0
            # Convert back to numpy for sounddevice
            audio_float = audio_float.cpu().numpy()
        else:
            # Original code
            audio_data = np.frombuffer(audio_buffer, dtype=np.int16)
            audio_float = audio_data.astype(np.float32) / 32767.0
        
        # Play the audio with proper device selection and error handling
        sd.play(audio_float, SAMPLE_RATE)
        sd.wait()
    except Exception as e:
        print(f"Audio playback error: {e}")
```

## WAV File Processing

For the `stitch_wav_files` function, optimize the NumPy operations for MPS:

```python
def stitch_wav_files(input_files, output_file, crossfade_ms=50):
    # Existing code...
    
    # Use torch for processing if MPS is available
    if MPS_AVAILABLE:
        # Initialize with empty tensor
        final_audio = torch.tensor([], dtype=torch.int16, device=mps_device)
        
        for i, input_file in enumerate(input_files):
            try:
                with wave.open(input_file, 'rb') as wav:
                    # Store params for first file
                    if first_params is None:
                        first_params = wav.getparams()
                    elif wav.getparams() != first_params:
                        print(f"Warning: WAV file {input_file} has different parameters")
                        
                    frames = wav.readframes(wav.getnframes())
                    # Convert to torch tensor on MPS
                    audio = torch.frombuffer(frames, dtype=torch.int16).to(mps_device)
                    
                    if i == 0:
                        # First segment - use as is
                        final_audio = audio
                    else:
                        # Apply crossfade with previous segment
                        if len(final_audio) >= crossfade_samples and len(audio) >= crossfade_samples:
                            # Create crossfade weights
                            fade_out = torch.linspace(1.0, 0.0, crossfade_samples, device=mps_device)
                            fade_in = torch.linspace(0.0, 1.0, crossfade_samples, device=mps_device)
                            
                            # Apply crossfade
                            crossfade_region = (final_audio[-crossfade_samples:] * fade_out + 
                                               audio[:crossfade_samples] * fade_in).to(torch.int16)
                            
                            # Combine
                            final_audio = torch.cat([final_audio[:-crossfade_samples], 
                                                    crossfade_region, 
                                                    audio[crossfade_samples:]])
                        else:
                            # One segment too short for crossfade, just append
                            print(f"Segment {i} too short for crossfade, concatenating directly")
                            final_audio = torch.cat([final_audio, audio])
            except Exception as e:
                print(f"Error processing file {input_file}: {e}")
                if i == 0:
                    raise  # Critical failure if first file fails
        
        # Move back to CPU for writing to file
        final_audio = final_audio.cpu().numpy()
    else:
        # Original NumPy implementation
        # ...
```

## Additional Optimizations

1. Enable mixed precision for MPS (available in macOS Sonoma and newer):

```python
# Add to the beginning of the file after hardware detection
USE_MIXED_PRECISION = False
if MPS_AVAILABLE:
    # Check macOS version for mixed precision support
    import platform
    mac_version = tuple(map(int, platform.mac_ver()[0].split('.')))
    if mac_version >= (14, 0):  # macOS Sonoma or newer
        USE_MIXED_PRECISION = True
        print("✓ Enabling mixed precision for MPS (bfloat16)")
```

2. Use autocast for mixed precision operations:

```python
# In functions that process tensors
if USE_MIXED_PRECISION and MPS_AVAILABLE:
    with torch.autocast(device_type="mps", dtype=torch.bfloat16):
        # Tensor operations here
```

These optimizations should help your code take advantage of Apple's MPS technology for improved performance on Apple Silicon Macs.

Citations:
[1] https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/8987996/0fd27449-ecd0-4050-bb70-15b0a74d637c/paste.txt
[2] https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/8987996/0fd27449-ecd0-4050-bb70-15b0a74d637c/paste.txt
[3] https://www.restack.io/p/ai-optimized-processors-answer-mps-features-cat-ai
[4] https://developer.apple.com/videos/play/tech-talks/111373/
[5] https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/ResourceOptions.html
[6] https://stackoverflow.com/questions/70240506/why-is-numpy-native-on-m1-max-greatly-slower-than-on-old-intel-i5
[7] https://discuss.pytorch.org/t/code-didnt-speed-up-as-expected-when-using-mps/152016
[8] https://huggingface.co/docs/diffusers/en/optimization/mps
[9] https://developer.apple.com/videos/play/wwdc2023/10050/
[10] https://developer.apple.com/metal/pytorch/
[11] https://developer.apple.com/videos/play/wwdc2021/10152/
[12] https://forums.developer.apple.com/forums/thread/112090
[13] https://developer.apple.com/documentation/xcode/optimizing-gpu-performance
[14] https://forums.developer.apple.com/forums/thread/111555
[15] https://developer.apple.com/documentation/metalperformanceshaders
[16] https://appleinsider.com/inside/xcode/tips/optimizing-metal-ways-to-improve-code-made-for-apples-graphics-framework
[17] https://discussions.apple.com/thread/254910760
[18] https://www.youtube.com/watch?v=LXTUFmbZwec
[19] https://developer.apple.com/videos/play/wwdc2020/10677/
[20] https://www.reddit.com/r/MachineLearning/comments/1gf46km/d_m4_chips_for_training_ml_mps/
[21] https://www.kodeco.com/books/metal-by-tutorials/v2.0/chapters/24-performance-optimization
[22] https://github.com/LaurentMazare/tch-rs/issues/676
[23] https://discuss.huggingface.co/t/mps-is-running-slower-than-cpu-on-mac-m1-pro/22828
[24] https://discourse.julialang.org/t/julia-access-to-apple-gpu-with-mlx-and-or-metal-performance-shaders-mps/117647
[25] https://developer.apple.com/videos/play/wwdc2024/10218/
[26] https://www.reddit.com/r/mac/comments/16j1b4h/python_run_faster_on_apple_silicon_than_the_x86/
[27] https://developer.apple.com/forums/tags/metal-performance-shaders
[28] https://developer.apple.com/documentation/apple-silicon/tuning-your-code-s-performance-for-apple-silicon
[29] https://www.reddit.com/r/LocalLLaMA/comments/1crwkia/why_people_buying_macs_instead_of_cuda_machines/
[30] https://github.com/Vaibhavs10/insanely-fast-whisper/issues/258
[31] https://stackoverflow.com/questions/77788295/is-there-a-better-way-to-optimize-metal-compute-shaders-that-operate-on-the-same
[32] https://www.reddit.com/r/macsysadmin/comments/13j8bqn/managing_our_only_appleenvironment_customer_best/
[33] https://developer.apple.com/videos/play/wwdc2016/606/
[34] https://forums.developer.apple.com/forums/thread/691917
[35] https://developer.apple.com/videos/play/wwdc2020/10677/?time=798
[36] https://stackoverflow.com/questions/79144698/pytorch-model-running-on-cpu-despite-mps-apple-silicon-being-available-and-det
[37] https://discuss.huggingface.co/t/autotrain-python-automatically-using-mps-on-mac-how-to-switch-to-cpu/116449

---
Answer from Perplexity: [pplx.ai/share](https://www.perplexity.ai/search/pplx.ai/share)

**Comment by Lex-au** [comment]
@dogjamboree Much appreciated! I will look into this on the weekend and see what I can do.

**Comment by jandrop** [comment]
Hi, there's any update on this?
